# 1. Présentation du modèle

In [59]:
import numpy as np
from movielensutils import *

## Question 1.1

In [7]:
R, mask = load_movielens("ml-100k/u.data")

In [8]:
print("Shape of R: ", R.shape)

('Shape of R: ', (943, 1682))


La matrice retrounée a bien la taille 943 x 1682.
L'option minidata permet de récupérer une sous-matrice de R de taille 100 x 200.

## Question 1.2

Il y 943 utilisateurs et 1682 films référencés.

In [9]:
print("Number of non-zeros", np.count_nonzero(mask))

('Number of non-zeros', 100000)


## Question 1.3

Pour $Q\in\mathcal{M_{\left|\mathrm{U}\right|\times\left|\mathrm{C}\right|}},$$P\in\mathcal{M_{\left|\mathrm{C}\right|\times\left|\mathrm{I}\right|}}$,
on a :

\[
\begin{array}{ccc}
f:(P,Q) & \mapsto & \frac{1}{2}{\displaystyle \sum_{(u,i)\in K}\left(r_{ui}-{\displaystyle \sum_{c\in C}q_{uc}p_{ci}}\right)^{2}+\frac{\rho}{2}\left(\sum_{u\in U,c\in C}q_{uc}^{2}+\sum_{c\in C,i\in I}p_{ci}^{2}\right)}\\
 & = & \frac{1}{2}{\displaystyle \sum_{(u,i)\in K}\sum_{c\in C}\left(\frac{r_{ui}}{\left|\mathrm{C}\right|}-{\displaystyle q_{uc}p_{ci}}\right)^{2}+\frac{\rho}{2}\left(\sum_{u\in U,c\in C}q_{uc}^{2}+\sum_{c\in C,i\in I}p_{ci}^{2}\right)}\\
 & = & \frac{1}{2}{\displaystyle \sum_{(u,i)\in K}\sum_{c\in C}\left(\frac{r_{ui}}{\left|\mathrm{C}\right|}^{2}-{\displaystyle 2\frac{r_{ui}}{\left|\mathrm{C}\right|}q_{uc}p_{ci}+\left(q_{uc}p_{ci}\right)^{2}}\right)+\frac{\rho}{2}\left(\sum_{u\in U,c\in C}q_{uc}^{2}+\sum_{c\in C,i\in I}p_{ci}^{2}\right)}\\
 & = & {\displaystyle \sum_{u\in U}\sum_{i\in I}\sum_{c\in C}\left(\frac{r_{ui}}{2\left|\mathrm{C}\right|}^{2}-{\displaystyle \frac{r_{ui}}{\left|\mathrm{C}\right|}q_{uc}p_{ci}}+\frac{\rho}{2}q_{uc}^{2}+\frac{\rho}{2}p_{ci}^{2}\right)}+{\displaystyle \sum_{u\in U}}{\displaystyle \sum_{i\in I}}{\displaystyle \sum_{c\in C}}\frac{1}{2}\left(q_{uc}p_{ci}\right)^{2}\\
 & = & {\displaystyle \sum_{u\in U}\sum_{i\in I}\sum_{c\in C}}\left(g\left(q_{uc},p_{ci}\right)+\frac{1}{2}h\left(q_{uc},p_{ci}\right)\right)
\end{array}
\]


où $g$ est une forme quadratique sur $\mathbb{R}^{2}$et $h:(x,y)\mapsto(xy)^{2}$

Pour $\overrightarrow{u}=(1,0),\overrightarrow{v}=(0,1),t=\frac{1}{2}$,
on a :

\begin{equation}
h\left(t\overrightarrow{u}+\left(1-t\right)\overrightarrow{v}\right)>th\left(\overrightarrow{u}\right)+\left(1-t\right)h\left(\overrightarrow{v}\right)
\end{equation}


$h$ est donc non convexe.

Par suite, $f$ est non convexe.
Pour $Q\in\mathcal{M_{\left|\mathrm{U}\right|\times\left|\mathrm{C}\right|}},$$P\in\mathcal{M_{\left|\mathrm{C}\right|\times\left|\mathrm{I}\right|}}$,
on a :
\begin{equation}
\overrightarrow{grad}f\left(P,Q\right)=\left(\begin{array}{c}
\frac{\partial f(P,Q)}{\partial P}\\
\frac{\partial f(P,Q)}{\partial Q}
\end{array}\right)
\end{equation}


et

\begin{equation}
f\left(P,Q\right)=\frac{1}{2}\left\Vert 1_{K}\circ\left(R-QP\right)\right\Vert _{F}^{2}+\frac{\rho}{2}\left(\left\Vert Q\right\Vert _{F}^{2}+\left\Vert P\right\Vert _{F}^{2}\right)
\end{equation}


On pose

\begin{equation}
\delta:(P,Q)\mapsto\left\Vert 1_{K}\circ\left(R-QP\right)\right\Vert _{F}^{2}
\end{equation}


et 
\begin{equation}
M=1_{K}\circ\left(R-QP\right)
\end{equation}


On obtient :

\begin{equation}
\begin{array}{ccc}
\partial M & = & \partial\left(1_{K}\circ\left(R-QP\right)\right)\\
 & = & \left(\partial1_{K}\right)\circ\left(R-QP\right)+1_{K}\circ\partial\left(R-QP\right)\\
 & = & 1_{K}\circ\left(\partial R-\partial\left(QP\right)\right)\\
 & = & -1_{K}\circ\left(\left(\partial Q\right)P+Q\left(\partial P\right)\right)
\end{array}
\end{equation}


D'où 

\begin{equation}
\begin{array}{ccc}
\frac{\partial f(P,Q)}{\partial P} & = & 2\left\langle M,\frac{\partial M}{\partial P}\right\rangle +\rho P\\
 & = & -Q^{t}\times1_{K}\circ\left(R-QP\right)+\rho P\\
\frac{\partial f(P,Q)}{\partial Q} & = & -1_{K}\circ\left(R-QP\right)\times P^{t}+\rho Q
\end{array}
\end{equation}


Les différentielles du gradient ne sont pas bornées. Donc il n'est
pas Lipschitzien.

# 2. Trouver $P$ lorsque $Q_{0}$ est fixé

In [10]:
from scipy.sparse.linalg import svds
C=6 #{0,...,6}
U, s, V = svds(R, k=C)
Q0=U[:,0:C]
P0=V[0:C,:]
print Q0.shape, P0.shape

rho=0.2

(943, 6) (6, 1682)


## Question 2.1

Pour $Q_{0}$ fixé, on a :
\begin{array}{ccc}
f(P,Q_{0}) & = & {\displaystyle \sum_{u\in U}\sum_{i\in I}\sum_{c\in C}\left(\frac{r_{ui}}{2\left|\mathrm{C}\right|}^{2}-{\displaystyle \frac{r_{ui}}{\left|\mathrm{C}\right|}q_{0uc}p_{ci}}+\frac{\rho}{2}q_{0uc}^{2}+\frac{\rho}{2}p_{ci}^{2}+\frac{1}{2}\left(q_{0uc}p_{ci}\right)^{2}\right)}\\
 & = & {\displaystyle \sum_{u\in U}\sum_{i\in I}\sum_{c\in C}}\gamma_{u,i,c}\left(p_{ci}\right)
\end{array}
 

avec $\gamma_{u,i,c}$
  forme quadratique sur $\mathbb{R}$
  de coefficient de plus haut degré positif. Il s'agit donc d'une forme convexe. D'où $f$
  est convexe.

Le gradient de f
  est 

$$\frac{\partial f(P,Q_0)}{\partial P}=-Q_0^{t}\times1_{K}\circ\left(R-Q_0P\right)+\rho P$$
 


## Question 2.2

In [13]:
from scipy.optimize import check_grad
def objective(P, Q0, R, mask, rho):
    """
    La fonction objectif du probleme simplifie.
    Prend en entree 
    P : la variable matricielle de taille C x I
    Q0 : une matrice de taille U x C
    R : une matrice de taille U x I
    mask : une matrice 0-1 de taille U x I
    rho : un reel positif ou nul

    Sorties :
    val : la valeur de la fonction
    grad_P : le gradient par rapport a P
    """

    tmp = (R - Q0.dot(P)) * mask

    val = np.sum(tmp ** 2)/2. + rho/2. * (np.sum(Q0 ** 2) + np.sum(P ** 2))

    grad_P = rho*P-(Q0.T).dot(tmp)

    return val, grad_P.ravel()

val, grad=objective(P0, Q0, R, mask, rho)

print check_grad(val,grad, P0.ravel())

TypeError: 'numpy.ndarray' object is not callable

## Question 2.3

In [34]:
def gradient(g, P0, gamma, epsilon):
    P=P0
    val, grad=g(P)
    nombre=0
    while(True):
        if np.linalg.norm(grad)<=epsilon:
            return val, P, nombre
        else:
            nombre=nombre+1
            P=P-gamma*grad.reshape(P.shape)
            val, grad=g(P)

## Question 2.4

In [37]:
epsilon=1.
cnstLips=rho+np.linalg.norm((Q0.T).dot(Q0))
gamma_donne=1./cnstLips
print("gamma utilisé:"+str(gamma_donne))
val,P_,nombre=gradient(g=lambda P:objective(P,Q0=Q0,R=R,mask=mask,rho=rho),P0=P0,gamma=gamma_donne,epsilon=epsilon)
print("valeur optimale: "+str(val)+"    nombre d'itération: "+str(nombre))

gamma utilisé:0.377431164897
valeur optimale: 303938.455544    nombre d'itération: 42


# Raffinements algorimtmiques pour le problème à $Q_{0}$ fixé

## Question 3.1

In [48]:
from scipy import optimize

def objectiveGamma(gamma,X0,direction,f):
    X=X0 + gamma*direction.reshape(X0.shape)
    val, grad_0 = f(X)
    
    return val


def gradient_LinearSearch(g, pointDepart, epsilon):
    pointCourant=pointDepart
    val, grad=g(pointCourant)
    nombre=0
    while(True):
        if np.linalg.norm(grad)<=epsilon:
            return val,pointCourant,nombre
        else:
            nombre=nombre+1
            direction=-grad
            gamma=optimize.brent(lambda gamma:objectiveGamma(gamma,X0=pointCourant,direction=direction,f=g))
            pointCourant=pointCourant+gamma*direction.reshape(pointCourant.shape)
            val, grad=g(pointCourant)

In [49]:
val,P_optimal,nombre=gradient_LinearSearch(g=lambda P:objective(P,Q0=Q0,R=R,mask=mask,rho=rho),pointDepart=P0,epsilon=epsilon)
print("valeur optimale: "+str(val)+"    nombre d'itération: "+str(nombre))

valeur optimale: 303937.249943    nombre d'itération: 11


## Question 3.2

La fonction $f$ est une somme de formes quadratiques sur $\mathbb{R}$ de coefficients de plus haut degré strictement positifs. De plus, la dimension de l'espace de départ est finie. Il est donc justifiable d'avoir recours à l'algorithme de gradient conjugué

In [55]:
def gradient_Conjuge(g, pointDepart, epsilon):
    pointCourant=pointDepart
    val, grad=g(pointCourant)
    direction=-grad
    nombre=0
    while(True):
        if np.linalg.norm(grad)<=epsilon:
            return val, pointCourant, nombre
        else:
            nombre=nombre+1
            gamma=optimize.brent(lambda gamma:objectiveGamma(gamma,X0=pointCourant,direction=direction,f=g))
            pointCourant=pointCourant+gamma*direction.reshape(pointCourant.shape)
            norm1=np.linalg.norm(grad)**2
            val, grad=g(pointCourant)
            norm2=np.linalg.norm(grad)**2
            b=norm2/norm1
            direction=-grad+b*direction.reshape(grad.shape)

In [56]:
val,P_optimal,nombre=gradient_Conjuge(g=lambda P:objective(P,Q0=Q0,R=R,mask=mask,rho=rho),pointDepart=P0,epsilon=epsilon)
print("valeur optimale: "+str(val)+"    nombre d'itération: "+str(nombre))

valeur optimale: 303936.977522    nombre d'itération: 7


## Question 3.3

In [57]:
import time
t0=time.clock()

print "Gradient with constant step size:"
t1=time.clock()
val,P_,nombre=gradient(g=lambda P:objective(P,Q0=Q0,R=R,mask=mask,rho=rho),P0=P0,gamma=gamma_donne,epsilon=epsilon)
t2=time.clock()
duree=t2-t1
print("value_optimal:"+str(val))
print("nombreIteration:"+str(nombre))
print("run time:"+str(duree))

print("")
print "Gradient with line search:"
t1=time.clock()
val,P_optimal,nombre=gradient_LinearSearch(g=lambda P:objective(P,Q0=Q0,R=R,mask=mask,rho=rho),pointDepart=P0,epsilon=epsilon)
t2=time.clock()
duree=t2-t1
print("value_optimal:"+str(val))
print("nombreIteration:"+str(nombre))
print("run time:"+str(duree))

print("")
print "Gradient conjugé:"
t1=time.clock()
val,P_optimal,nombre=gradient_Conjuge(g=lambda P:objective(P,Q0=Q0,R=R,mask=mask,rho=rho),pointDepart=P0,epsilon=epsilon)
t2=time.clock()
duree=t2-t1
print("value_optimal:"+str(val))
print("nombreIteration:"+str(nombre))
print("run time:"+str(duree))

Gradient with constant step size:
value_optimal:303938.455544
nombreIteration:42
run time:4.058878

Gradient with line search:
value_optimal:303937.249943
nombreIteration:11
run time:23.115161

Gradient conjugé:
value_optimal:303936.977522
nombreIteration:7
run time:12.857316


# 4 Résolution du problème complet

## Question 4.1

In [70]:
def total_objective(P, Q, R, mask, rho):
    """
    La fonction objectif du probleme complet.
    Prend en entree 
    P : la variable matricielle de taille C x I
    Q : la variable matricielle de taille U x C
    R : une matrice de taille U x I
    mask : une matrice 0-1 de taille U x I
    rho : un reel positif ou nul

    Sorties :
    val : la valeur de la fonction
    grad_P : le gradient par rapport a P
    grad_Q : le gradient par rapport a Q
    """

    tmp = (R - Q.dot(P)) * mask

    val = np.sum(tmp ** 2)/2. + rho/2. * (np.sum(Q ** 2) + np.sum(P ** 2))

    grad_P = rho*P-(Q.T).dot(tmp)

    grad_Q = rho*Q-tmp.dot(P.T)

    return val, grad_P, grad_Q
def total_objective_vectorized(PQvec, R, mask, rho):
    """
    Vectorisation de la fonction precedente de maniere a ne pas
    recoder la fonction gradient
    """

    # reconstruction de P et Q
    n_items = R.shape[1]
    n_users = R.shape[0]
    F = PQvec.shape[0] / (n_items + n_users)
    Pvec = PQvec[0:n_items*F]
    Qvec = PQvec[n_items*F:]
    P = np.reshape(Pvec, (F, n_items))
    Q = np.reshape(Qvec, (n_users, F))

    val, grad_P, grad_Q = total_objective(P, Q, R, mask, rho)
    return val, np.concatenate([grad_P.ravel(), grad_Q.ravel()])


In [ ]:
PQ0=np.concatenate([P0.ravel(), Q0.ravel()])
val, PQ, nombre = gradient_LinearSearch(lambda PQvec:total_objective_vectorized(PQvec, R=R, mask=mask, rho=rho),pointDepart=PQ0,epsilon=100)
print("valeur optimale: "+str(val)+"    nombre d'itération: "+str(nombre))

## Question 4.2

Pour $Q_{k-1}$
 fixé, on a :

$$f\left(P_{k},Q_{k-1}\right)\leq f\left(P_{k-1},Q_{k-1}\right)$$
 

D'autre part, on a :

\begin{array}{cc}
 & f\left(P_{k},Q_{k}\right)\leq f\left(P,Q_{k-1}\right)\\
\Leftrightarrow & f\left(P_{k},Q_{k}\right)\leq f\left(P_{k},Q_{k-1}\right)
\end{array}
 

D'après (1) et (2), on a $$f\left(P_{k},Q_{k}\right)\leq f\left(P_{k-1},Q_{k-1}\right)$$
 

Ainsi la suite réelle $\left(u_{k}\right)_{k}=\left(f\left(P_{k},Q_{k}\right)\right)_{k}$
  est positive et décroissante. Elle est donc convergente 

## Question 4.3

In [61]:
def objective_Q(Q, P0, R, mask, rho):
    val, grad_P, grad_Q=total_objective(P=P0,Q=Q,R=R,mask=mask,rho=rho)
    return val, grad_Q



def gradient_MoindreCarre(P0,Q0,R,mask,rho,epsilon):
    nombre=0
    val_avant=0
    P_courant=P0
    Q_courant=Q0

    val1,P_optimal,nbr=gradient_LinearSearch(g=lambda P:objective(P,Q0=Q_courant,R=R,mask=mask,rho=rho),pointDepart=P_courant,epsilon=epsilon)
    nombre=nombre+nbr
    P_courant=P_optimal

    val2, Q_optimal,nbr= gradient_LinearSearch(g=lambda Q:objective_Q(Q,P0=P_courant,R=R,mask=mask,rho=rho),pointDepart=Q_courant,epsilon=epsilon)
    nombre=nombre+nbr
    Q_courant=Q_optimal

    difference=np.abs(val1-val2)

    while(True):
        if difference<=epsilon :
            return val2,P_courant,Q_courant,nombre
        else:
            val1,P_optimal,nbr=gradient_LinearSearch(g=lambda P:objective(P,Q0=Q_courant,R=R,mask=mask,rho=rho),pointDepart=P_courant,epsilon=epsilon)
            nombre=nombre+nbr
            P_courant=P_optimal

            val2, Q_optimal,nbr= gradient_LinearSearch(g=lambda Q:objective_Q(Q,P0=P_courant,R=R,mask=mask,rho=rho),pointDepart=Q_courant,epsilon=epsilon)
            nombre=nombre+nbr
            Q_courant=Q_optimal

            difference=np.abs(val1-val2)

In [ ]:
val, P, Q, nombre=gradient_MoindreCarre(P0=P0,Q0=Q0,R=R,mask=mask,rho=rho,epsilon=100)
print("valeur optimale: "+str(val)+"    nombre d'itération: "+str(nombre))

## Question 4.4

In [ ]:
import time
t0=time.clock()

print "Gradient with constant step size:"
t1=time.clock()
val,P_,nombre=gradient(g=lambda P:objective(P,Q0=Q0,R=R,mask=mask,rho=rho),P0=P0,gamma=gamma_donne,epsilon=epsilon)
t2=time.clock()
duree=t2-t1
print("value_optimal:"+str(val))
print("nombreIteration:"+str(nombre))
print("run time:"+str(duree))

print("")
print "Gradient with line search:"
t1=time.clock()
val,P_optimal,nombre=gradient_LinearSearch(g=lambda P:objective(P,Q0=Q0,R=R,mask=mask,rho=rho),pointDepart=P0,epsilon=epsilon)
t2=time.clock()
duree=t2-t1
print("value_optimal:"+str(val))
print("nombreIteration:"+str(nombre))
print("run time:"+str(duree))